data/blog/enjoyable-database-testing.md

```diff
@@ -0,0 +1,277 @@
+You've written some Haskell code and it compiles ... so it works?
+
+Hahahahahahahaha ha ... sigh ... *[muffled sobbing]*.
+
+Turns out this is not the case. Not the case in general and definitely not the case if you are writing database queries.
+
+You can try to find various libraries to limit what needs to be tested but at some point you are going to want to test your database code.
+
+`tmp-postgres` can help you write reliable tests. There are a lot of ways you could utilize `tmp-postgres` to write your tests but I'll show you what I find works best for me.
+
+The blog post also shows general database tests good practices that are not tied to using `tmp-postgres` or `postgresql-simple`.
+
+## Starting `postgres` Quickly
+
+We need to create a fast `tmp-postgres` setup function.
+
+First will utilize `initdb` caching by using [`withDbCache`](https://hackage.haskell.org/package/tmp-postgres-1.34.0.0/docs/Database-Postgres-Temp.html#v:withDbCache). As I discussed [previously](/faster-database-testing.html)
+this gives a 3-4x performance boost. However in "real" projects the overhead in database testing tends to come from the time it takes to create a migrated database. Running a complete set of database migration can easily take 10 seconds.
+To speed up this process we need a way to cache a database cluster after the migrations have run.
+
+`tmp-postgres` provides the [`cacheAction`](https://hackage.haskell.org/package/tmp-postgres-1.34.0.0/docs/Database-Postgres-Temp.html#v:cacheAction) function to cache migrated database clusters. Here is the type signature:
+
+```haskell
+cacheAction :: FilePath -> (DB -> IO ()) -> Config -> IO (Either StartError Config)
+```
+
+If database cluster folder (the first argument) does not exist, the continuation (second argument) will run. The third argument is to configure the temporary database which makes the cluster. `cacheAction` returns a Config that can be used to start a database initialized with the cached database cluster referred to in the first argument.
+
+Long story short you should use `cacheAction` to store a database cluster at the state after the migration has been run stored at a location based on the hash of the migration query and (any other previous hashes ... if you wanted to cache every migration for instance). If you do this you won't have to run your migrations every time you run your tests.
```

Comment:
Long story short, you should use `cacheAction` to store a database cluster at the state after the migration has been run, and stored at a location based on the hash of the migration query (and any other previous hashes ... if you wanted to cache every migration for instance).